File: backend/src/test/java/ca/metricalsky/winston/repository/AuthorRepositoryTest.java
package ca.metricalsky.winston.repository;

import ca.metricalsky.winston.entity.AuthorEntity;
import ca.metricalsky.winston.test.TestUtils;
import ca.metricalsky.winston.test.annotations.RepositoryTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import static org.assertj.core.api.Assertions.assertThat;

@RepositoryTest
class AuthorRepositoryTest {

    private static final String DISPLAY_NAME = "@displayName";
    private static final String CHANNEL_URL = "http://www.youtube.com/c/channelUrl";

    @Autowired
    private TestEntityManager entityManager;
    @Autowired
    private AuthorRepository repository;

    private AuthorEntity savedAuthor;

    @BeforeEach
    void beforeEach() {
        savedAuthor = entityManager.persist(AuthorEntity.builder()
                .id(TestUtils.randomId())
                .displayName(DISPLAY_NAME)
                .channelUrl(CHANNEL_URL)
                .build());
    }

    @Test
    void findByChannelUrl() {
        var author = repository.findByChannelUrl(CHANNEL_URL);

        assertThat(author)
                .contains(savedAuthor);
    }

    @Test
    void findByChannelUrl_notFound() {
        var author = repository.findByChannelUrl("");

        assertThat(author)
                .isEmpty();
    }

    @Test
    void findByDisplayName() {
        var author = repository.findByDisplayName(DISPLAY_NAME);

        assertThat(author)
                .contains(savedAuthor);
    }

    @Test
    void findByDisplayName_notFound() {
        var author = repository.findByDisplayName("");

        assertThat(author)
                .isEmpty();
    }
}

File: database/src/test/kotlin/ca/metricalsky/winston/database/repository/author/AuthorRepositoryTest.kt
package ca.metricalsky.winston.database.repository.author

import ca.metricalsky.winston.database.entity.author.AuthorEntity
import ca.metricalsky.winston.database.test.annotation.DatabaseTest
import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager

@DatabaseTest
class AuthorRepositoryTest(
    authorRepository: AuthorRepository,
    entityManager: TestEntityManager,
): WordSpec({

    val faker = DatabaseFaker()

    var author: AuthorEntity? = null

    beforeEach {
        author = entityManager.persist(faker.author().completeEntity())
    }

    "findByChannelUrl()" should {

        "return an author when given its channel URL" {
            val result = authorRepository.findByChannelUrl(author?.channelUrl!!)
            result shouldBe author
        }

        "return null when given an empty channel URL" {
            val result = authorRepository.findByChannelUrl("")
            result shouldBe null
        }
    }

    "findByDisplayName()" should {

        "return an author when given its display name" {
            val result = authorRepository.findByDisplayName(author?.displayName!!)
            result shouldBe author
        }

        "return null when given an empty display name" {
            val result = authorRepository.findByDisplayName("")
            result shouldBe null
        }
    }
})
